File: sources/app/src/androidTest/java/io/github/sh4/zabuton/LibGit2Test.java
package io.github.sh4.zabuton;

import android.content.Context;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.platform.app.InstrumentationRegistry;

import org.apache.commons.io.FileUtils;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Comparator;

import io.github.sh4.zabuton.git.ICloneProgress;
import io.github.sh4.zabuton.git.IFetchProgress;
import io.github.sh4.zabuton.git.LibGit2;
import io.github.sh4.zabuton.git.Repository;
import io.github.sh4.zabuton.git.ResetKind;
import io.github.sh4.zabuton.util.ContextUtil;

import static org.junit.Assert.*;

@RunWith(AndroidJUnit4.class)
public class LibGit2Test {
    static {
        System.loadLibrary("native-lib");
    }

    private static final String CLONE_URL = "https://github.com/sh4/test-git.git";

    @Test
    public void cloneAndOpen() throws IOException {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        Path reposPath = context.getDir("test-repos", Context.MODE_PRIVATE).toPath();
        cleanupRepos(reposPath);
        initLibGit2(context);
        final ICloneProgress[] p = {null};
        assertNotNull(Repository.clone(CLONE_URL, reposPath.toString(), cur -> p[0] = cur));
        assertNotEquals(0, p[0].getCompletedSteps());
        assertNotEquals(0, p[0].getTotalSteps());
        assertNotEquals(0, p[0].getIndexedDeltas());
        assertNotEquals(0, p[0].getIndexedObjects());
        assertEquals(0, p[0].getLocalObjects()); // Local object is 0 because it is clone to empty repository.
        assertNotEquals(0, p[0].getReceivedBytes());
        assertNotEquals(0, p[0].getReceivedObjects());
        assertNotEquals(0, p[0].getTotalDeltas());
        assertNotEquals(0, p[0].getTotalObjects());
        assertNotEquals(1, Files.walk(reposPath).count());

        assertNotNull(Repository.open(reposPath.toString()));
    }

    @Test
    public void fetchRemote() {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        initLibGit2(context);
        Path reposPath = context.getDir("test-repos", Context.MODE_PRIVATE).toPath();
        Repository repos = ensureRepositoryOpened(reposPath);
        assertNotNull(repos);
        final IFetchProgress[] p = {null};
        repos.fetch("origin", cur -> p[0] = cur);
        assertNotNull(p[0]);
        //assertNotEquals(0, p[0].getReceivedBytes());
    }

    private Repository ensureRepositoryOpened(Path reposPath) {
        return Files.exists(reposPath)
                ? Repository.open(reposPath.toString())
                : Repository.clone(CLONE_URL, reposPath.toString(), cur -> {});
    }

    @Test
    public void checkoutBranch()  throws IOException {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        initLibGit2(context);
        Path reposPath = context.getDir("test-repos", Context.MODE_PRIVATE).toPath();
        cleanupRepos(reposPath);
        Repository repos = Repository.clone(CLONE_URL, reposPath.toString(), cur -> {});
        repos.fetch("origin", cur -> {});

        File anotherBranchFile = new File(reposPath.toFile(), "Test/Another.txt");
        assertFalse(anotherBranchFile.exists());

        repos.checkout("origin/sh4-patch-1", p -> {});
        assertTrue(anotherBranchFile.exists());

        repos.checkout("master", p -> {});
        assertFalse(anotherBranchFile.exists());
    }

    @Test
    public void reset() throws IOException {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        initLibGit2(context);
        Path reposPath = context.getDir("test-repos", Context.MODE_PRIVATE).toPath();
        File testFile = new File(reposPath.toFile(), "Test/Files.txt");

        FileUtils.writeStringToFile(testFile, "Foobar2000", "utf-8");
        assertEquals("Foobar2000", FileUtils.readFileToString(testFile, "utf-8"));

        Repository repos = ensureRepositoryOpened(reposPath);
        repos.reset(ResetKind.HARD, p -> {});

        assertNotEquals("Foobar2000", FileUtils.readFileToString(testFile, "utf-8"));
    }

    @Test
    public void branchOperations() throws IOException {
        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
        initLibGit2(context);
        Path reposPath = context.getDir("test-repos", Context.MODE_PRIVATE).toPath();
        cleanupRepos(reposPath);
        Repository repos = ensureRepositoryOpened(reposPath);

        String branch;
        branch = repos.getHeadName();
        assertEquals("master", branch);
        repos.checkout("origin/sh4-patch-1", p -> {});

        branch = repos.getHeadName();
        assertEquals("sh4-patch-1", repos.getHeadName());
        repos.checkout("origin/master", p -> {});

        branch = repos.getHeadName();
        assertEquals("master", repos.getHeadName());

        {
            String[] names = repos.getLocalBranchNames();
            assertArrayEquals(new String[] {"master", "sh4-patch-1"}, names);
        }

        {
            String[] names = repos.getRemoteBranchNames();
            assertArrayEquals(new String[] {"origin/master", "origin/sh4-patch-1"}, names);
        }
    }

    private void cleanupRepos(Path reposPath) throws IOException {
        if (Files.exists(reposPath)) {
            Files.walk(reposPath)
                    .sorted(Comparator.reverseOrder())
                    .filter(x -> !reposPath.equals(x))
                    .map(Path::toFile)
                    .forEach(File::delete);
            Files.delete(reposPath);
        }
    }

    private void initLibGit2(Context context) {
        File sslCertificatesFile = ContextUtil.writeToFilesDir(context, "build/cacert.pem");
        assertNotNull(sslCertificatesFile);
        LibGit2.init(sslCertificatesFile.getAbsolutePath());
    }
}

File: sources/app/src/androidTest/java/io/github/sh4/zabuton/LibGit2Test.kt
package io.github.sh4.zabuton

import android.content.Context
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import io.github.sh4.zabuton.git.*
import io.github.sh4.zabuton.workspace.initializeLibGit2
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import java.io.File
import java.io.IOException
import java.nio.charset.Charset

@RunWith(AndroidJUnit4::class)
class LibGit2Test {
    companion object {
        private const val CLONE_URL = "https://github.com/sh4/test-git.git"

        init {
            System.loadLibrary("native-lib")
        }
    }

    @Test
    @Throws(IOException::class)
    fun cloneAndOpen() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        val reposPath = context.getDir("test-repos", Context.MODE_PRIVATE)
        reposPath.deleteRecursively()
        initializeLibGit2(context)
        val p = arrayOf<ICloneProgress?>(null)
        Assert.assertNotNull(Repository.clone(CLONE_URL, reposPath.absolutePath) { cur: ICloneProgress? -> p[0] = cur })
        Assert.assertNotEquals(0, p[0]!!.completedSteps)
        Assert.assertNotEquals(0, p[0]!!.totalSteps)
        Assert.assertNotEquals(0, p[0]!!.indexedDeltas)
        Assert.assertNotEquals(0, p[0]!!.indexedObjects)
        Assert.assertEquals(0, p[0]!!.localObjects) // Local object is 0 because it is clone to empty repository.
        Assert.assertNotEquals(0, p[0]!!.receivedBytes)
        Assert.assertNotEquals(0, p[0]!!.receivedObjects)
        Assert.assertNotEquals(0, p[0]!!.totalDeltas)
        Assert.assertNotEquals(0, p[0]!!.totalObjects)
        Assert.assertNotEquals(1, reposPath.walk().count())
        Assert.assertNotNull(Repository.open(reposPath.absolutePath))
    }

    @Test
    fun fetchRemote() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        initializeLibGit2(context)
        val reposPath = context.getDir("test-repos", Context.MODE_PRIVATE)
        val repos = ensureRepositoryOpened(reposPath)
        Assert.assertNotNull(repos)
        val p = arrayOf<IFetchProgress?>(null)
        repos.fetch("origin") { cur: IFetchProgress? -> p[0] = cur }
        Assert.assertNotNull(p[0])
        //assertNotEquals(0, p[0].getReceivedBytes());
    }

    private fun ensureRepositoryOpened(reposPath: File): Repository {
        if (reposPath.exists()) {
            return Repository.open(reposPath.absolutePath)
        } else {
            return Repository.clone(CLONE_URL, reposPath.absolutePath) { cur: ICloneProgress? -> }
        }
    }

    @Test
    @Throws(IOException::class)
    fun checkoutBranch() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        initializeLibGit2(context)
        val reposPath = context.getDir("test-repos", Context.MODE_PRIVATE)
        reposPath.deleteRecursively()
        val repos = Repository.clone(CLONE_URL, reposPath.absolutePath) { cur: ICloneProgress? -> }
        repos.fetch("origin") { cur: IFetchProgress? -> }
        val anotherBranchFile = File(reposPath, "Test/Another.txt")
        Assert.assertFalse(anotherBranchFile.exists())
        repos.checkout("origin/sh4-patch-1") { p: ICheckoutProgress? -> }
        Assert.assertTrue(anotherBranchFile.exists())
        repos.checkout("master") { p: ICheckoutProgress? -> }
        Assert.assertFalse(anotherBranchFile.exists())
    }

    @Test
    @Throws(IOException::class)
    fun reset() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        initializeLibGit2(context)
        val reposPath = context.getDir("test-repos", Context.MODE_PRIVATE)
        val testFile = File(reposPath, "Test/Files.txt")
        testFile.writeText("Foobar2000", Charset.defaultCharset());
        Assert.assertEquals("Foobar2000", testFile.readText(Charset.defaultCharset()))
        val repos = ensureRepositoryOpened(reposPath)
        repos.reset(ResetKind.HARD) { p: ICheckoutProgress? -> }
        Assert.assertNotEquals("Foobar2000", testFile.readText(Charset.defaultCharset()))
    }

    @Test
    @Throws(IOException::class)
    fun branchOperations() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        initializeLibGit2(context)
        val reposPath = context.getDir("test-repos", Context.MODE_PRIVATE)
        reposPath.deleteRecursively()
        val repos = ensureRepositoryOpened(reposPath)
        var branch: String?
        branch = repos.headName
        Assert.assertEquals("master", branch)
        repos.checkout("origin/sh4-patch-1") { p: ICheckoutProgress? -> }
        branch = repos.headName
        Assert.assertEquals("sh4-patch-1", repos.headName)
        repos.checkout("origin/master") { p: ICheckoutProgress? -> }
        branch = repos.headName
        Assert.assertEquals("master", repos.headName)
        run {
            val names = repos.localBranchNames
            Assert.assertArrayEquals(arrayOf("master", "sh4-patch-1"), names)
        }
        run {
            val names = repos.remoteBranchNames
            Assert.assertArrayEquals(arrayOf("origin/master", "origin/sh4-patch-1"), names)
        }
    }

    @Test
    fun tags() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        initializeLibGit2(context)
        val reposPath = context.getDir("test-repos", Context.MODE_PRIVATE)
        reposPath.deleteRecursively()
        val repos = ensureRepositoryOpened(reposPath)
        val tags = repos.tagNames
        Assert.assertTrue(tags.isNotEmpty())
    }

    @Test
    fun logs() {
        val context = InstrumentationRegistry.getInstrumentation().targetContext
        initializeLibGit2(context)
        val reposPath = context.getDir("test-repos", Context.MODE_PRIVATE)
        reposPath.deleteRecursively()
        val repos = ensureRepositoryOpened(reposPath)
        repos.log { commit ->
            Assert.assertNotNull(commit)
            return@log true
        }
    }
}